### ISSUE TITLE
cda-service overwrites query columns with same name

### ISSUE THREAD

**Issue by ianfore** [CancerDataAggregator/cda-service#67]
The following demonstrates how patient id becomes substituted with the subject id. The query includes both the 

```
two_id_query = '''select p.id ,  s1.id
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1,
    UNNEST(s1.Diagnosis) as d1
    where i1.system ='GDC' 
    order by p.id '''
api_instance.sql_query('v3',two_id_query,
                      limit=10)
```
gives
```[{'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad'},
  {'id': 'e6915db0-7c89-484d-8f9f-15cca68b82fc'},
  {'id': '16614d46-172b-479c-992b-e80a8e9a2c59'},
  {'id': '567fc9e3-17a6-42b1-a896-5e9a9507d1d8'},
  {'id': '54e89878-a1bc-4f5a-9d68-4842a469586e'},
  {'id': 'a1c7b7b9-b8c8-48c3-9420-55497f9318fd'},
  {'id': 'ce3c8b98-e275-4cfd-a379-940d675a564b'},
  {'id': 'e4ce89ef-bcaa-418a-8a6b-3602793b9bbf'},
  {'id': '19d3c861-8a5f-49a2-acc0-b55b25465c35'},
  {'id': 'afae8dce-294a-4108-bb28-376f804ae5c4'}]
```

Only a single column is shown in the results. If a query like this is run directly in the BigQuery API then two columns are returned. The second column is automatically named 'id_1' by BigQuery.

It seems that the CDA parses the BigQuery results and treats both columns as id, overwriting p.id with s.id.

This likely explains why the patient id contained the subject id in https://github.com/CancerDataAggregator/cda-python/issues/16

The following confirms that the problem occurs for another query showing that it misses other columns with the same name

```
select p.id ,  su.id, su_id.system, su_id.value, sp_id.system, sp_id.value
    FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as su,
    UNNEST(su.Identifier) as su_id,
    unnest(Specimen) as sp,
    UNNEST(sp.Identifier) as sp_id
    order by p.id
```

returns only one each of system and value

```
[{'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '41d6aa10-181a-47cc-8eb5-3a28a8421a7d'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '65be1f28-b428-59e5-a187-22335ec57c17'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '786c67dd-7425-5804-97ce-7b64e96569c1'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '49602e48-0ccf-5f3c-9529-255fc3f3368f'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '7c2a9d5c-2315-57bd-ade5-e0932b421b0e'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': 'e0c6965b-fe8a-4ada-bf8f-c6a00d506ecd'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '721645ba-916b-5481-bf76-63891cb79836'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': 'dcc36b6e-7743-40b4-98eb-06580a694e0e'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '3ae5c716-7061-4236-8a90-bddc4e6e4834'},
  {'id': 'be37f1f7-2f98-4f74-bc04-6dd2ae2afcad',
   'system': 'GDC',
   'value': '14fbb898-2e88-5dd5-b7f6-6783a89754f4'}]
```

**Comment by pshapiro4broad** [comment]
I can reproduce this problem. It looks like the way we're converting the query results to JSON (by using `SELECT TO_JSON_STRING`) doesn't perform the uniquifying operation that the BQ console does. For now, you'll need to work around it by naming the selected values using `AS`, e.g.:
```sql
select p.id as pid,  s1.id as sid FROM gdc-bq-sample.cda_mvp.v3 p,
    unnest(ResearchSubject) as s1,
    UNNEST(s1.identifier) as i1,
    UNNEST(s1.Diagnosis) as d1
    where i1.system ='GDC' 
    order by p.id
```

**Comment by pshapiro4broad** [comment]
Another sql query to use to test
```sql
SELECT
  p.id,
  rs.id,
  sp.id
FROM
  `gdc-bq-sample.cda_mvp.v3` AS p,
  UNNEST(ResearchSubject) AS rs,
  UNNEST(rs.Specimen) as sp
WHERE
  p.id = 'TARGET-40-NAAWES'
```

**Comment by pshapiro4broad** [comment]
Interestingly, the duplicate keys are in the JSON response. But most JSON parsers don't support identical keys (since they're disallowed in the JSON spec) so only the first key is preserved. You can see this by running the query directly using `curl`:
```bash
$ curl -X POST "https://cda.cda-dev.broadinstitute.org/api/v1/sql-query/v3" -H "accept: application/json" -H "Content-Type: text/plain" -d "SELECT p.id, rs.id, sp.id FROM gdc-bq-sample.cda_mvp.v3 AS p, UNNEST(ResearchSubject) AS rs, UNNEST(rs.Specimen) as sp WHERE p.id = 'TARGET-40-NAAWES'"
```

**Comment by ianfore** [comment]
Interesting. See [this notebook](https://github.com/ianfore/cdatest/blob/main/Column%20name%20parsing.ipynb) which explores these queries. I added a raw query against the API which showed that the Python requests module returns the duplicates in the raw content, but identified an issue .json() function which loses them.

Not familiar with the TO_JSON_STRING function and how it behaves. I've just been querying BigQuery via its API and converting to json afterwards. As in the UI, the uniquifying behavior is present in the API.

Suggested two issues for the CDA API
It would be preferable if the API validates that it is returning legal json.
That the API passes through unique column names

Was aware of the workaround of providing aliases for the columns but felt it was important to be aware of what the REST API was doing. That may have other implications.

**Comment by pshapiro4broad** [comment]
I'm going to fix this anyway, but I wonder what the utility of supporting ambiguous names is. When this is fixed the query will return id, id_1, id_2. How will a user know which id is which? If they really want different columns it seems like they'd want them to have identifiable names in the result.

**Comment by ianfore** [comment]
Thanks - if a fix is straightforward it seems worth doing. I think asking that the api returns legal json is reasonable. If it were a lot of work it's less clear and not fixing would be reasonable. It seemed to me it at least needed thinking through to assess impact/risk. In a time crunch an issue like the [same patient being returned more than once](https://github.com/CancerDataAggregator/cda-service/issues/69) seems more worthy of effort 

It's not a unique problem. requests.Response.json() and even the json module in Python don't raise an exception on duplicate keys. It's a known behavior [see here](https://docs.python.org/3/library/json.html#repeated-names-within-an-object) where they effectively punted the problem. I worked this through in an [updated version of the notebook](https://github.com/ianfore/cdatest/blob/main/Column%20name%20parsing.ipynb) I shared earlier. 

Re the ambiguous names. The user creates the ambiguity when they write the query and they would have to solve what they created :-) The API wouldn't as such be creating the issue - just return the ambiguity they created Agree they should give them unique names.
